common/types/encoding/codecv1/codecv1.go

```diff
@@ -0,0 +1,497 @@
+					continue
+				}
+				currentIndex := tx.Nonce
+
+				if currentIndex < nextIndex {
+					return nil, 0, fmt.Errorf("unexpected batch payload, expected queue index: %d, got: %d. Batch index: %d, chunk index in batch: %d, block index in chunk: %d, block hash: %v, transaction hash: %v", nextIndex, currentIndex, batchIndex, chunkID, blockID, block.Header.Hash(), tx.TxHash)
+				}
+
+				// mark skipped messages
+				for skippedIndex := nextIndex; skippedIndex < currentIndex; skippedIndex++ {
+					quo := int((skippedIndex - baseIndex) / 256)
+					rem := int((skippedIndex - baseIndex) % 256)
+					for len(skippedBitmap) <= quo {
+						bitmap := big.NewInt(0)
+						skippedBitmap = append(skippedBitmap, bitmap)
+					}
+					skippedBitmap[quo].SetBit(skippedBitmap[quo], rem, 1)
+				}
+
+				// process included message
+				quo := int((currentIndex - baseIndex) / 256)
+				for len(skippedBitmap) <= quo {
+					bitmap := big.NewInt(0)
+					skippedBitmap = append(skippedBitmap, bitmap)
+				}
+
+				nextIndex = currentIndex + 1
+			}
+		}
+	}
+
+	bitmapBytes := make([]byte, len(skippedBitmap)*32)
+	for ii, num := range skippedBitmap {
+		bytes := num.Bytes()
+		padding := 32 - len(bytes)
+		copy(bitmapBytes[32*ii+padding:], bytes)
+	}
+
+	return bitmapBytes, nextIndex, nil
+}
+
+// constructSkippedBitmap constructs the 4844 blob payload.
+func constructBlobPayload(chunks []*encoding.Chunk) (*kzg4844.Blob, *kzg4844.Point, error) {
+	// the raw (un-padded) blob payload
+	blobBytes := make([]byte, 2*31)
+
+	// the number of chunks that contain at least one L2 transaction
+	numNonEmptyChunks := 0
+
+	// challenge digest preimage
+	// 1 hash for metadata and 1 for each chunk
+	challengePreimage := make([]byte, 16*32)
+
+	// the challenge point z
+	var z kzg4844.Point
+
+	// encode blob metadata and L2 transactions,
+	// and simultaneously also build challenge preimage
+	for chunkID, chunk := range chunks {
+		currentChunkStartIndex := len(blobBytes)
+		hasL2Tx := false
+
+		for _, block := range chunk.Blocks {
+			for _, tx := range block.Transactions {
+				if tx.Type != types.L1MessageTxType {
+					hasL2Tx = true
+					// encode L2 txs into blob payload
+					rlpTxData, err := encoding.ConvertTxDataToRLPEncoding(tx)
+					if err != nil {
+						return nil, nil, err
+					}
+					blobBytes = append(blobBytes, rlpTxData...)
+					continue
+				}
+			}
+		}
+
+		// blob metadata: chunki_size
+		chunkSize := len(blobBytes) - currentChunkStartIndex
+		binary.BigEndian.PutUint32(blobBytes[2+4*chunkID:], uint32(chunkSize))
+
+		if hasL2Tx {
+			numNonEmptyChunks++
+		}
+
+		// challenge: compute chunk data hash
+		hash := crypto.Keccak256Hash(blobBytes[currentChunkStartIndex:])
+		copy(challengePreimage[32+chunkID*32:], hash[:])
+	}
+
+	// blob metadata: num_chunks
+	binary.BigEndian.PutUint16(blobBytes[0:], uint16(numNonEmptyChunks))
+
+	// challenge: compute metadata hash
+	hash := crypto.Keccak256Hash(blobBytes[0:62])
```

Comment:
Should we declare 62 as `MAX_NUM_CHUNKS * 4 + 2`?

Reply:
Done in [f158118](https://github.com/scroll-tech/scroll/pull/1193/commits/f1581185883bb182ce3424b2157935693abd9bda).